src/sqlfluff/dialects/dialect_tsql.py

```diff
@@ -659,3 +661,101 @@ class OverlapsClauseSegment(BaseSegment):
 
     type = "overlaps_clause"
     match_grammar = Nothing()
+
+
+@tsql_dialect.segment()
+class DateAddFunctionNameSegment(BaseSegment):
+    """DATEADD function name segment.
+
+    Need to be able to specify this as type function_name
+    so that linting rules identify it properly
+    """
+
+    type = "function_name"
+    match_grammar = Sequence("DATEADD")
+
+
+@tsql_dialect.segment(replace=True)
+class FunctionSegment(BaseSegment):
+    """A scalar or aggregate function.
+
+    Maybe in the future we should distinguish between
+    aggregate functions and other functions. For now
+    we treat them the same because they look the same
+    for our purposes.
+    """
+
+    type = "function"
+    match_grammar = OneOf(
+        Sequence(
+            Sequence(
+                Ref("DateAddFunctionNameSegment"),
+                Bracketed(
+                    Delimited(
+                        OneOf(
+                            "D",
+                            "DAY",
+                            "DAYOFYEAR",
+                            "DD",
+                            "DW",
+                            "DY",
+                            "HH",
+                            "HOUR",
+                            "M",
+                            "MCS",
+                            "MI",
+                            "MICROSECOND",
+                            "MILLISECOND",
+                            "MINUTE",
+                            "MM",
+                            "MONTH",
+                            "MS",
+                            "N",
+                            "NANOSECOND",
+                            "NS",
+                            "Q",
+                            "QQ",
+                            "QUARTER",
+                            "S",
+                            "SECOND",
+                            "SS",
+                            "W",
+                            "WEEK",
+                            "WEEKDAY",
+                            "WK",
+                            "WW",
+                            "YEAR",
+                            "Y",
+                            "YY",
+                            "YYYY",
+                        ),
+                        Ref(
+                            "FunctionContentsGrammar",
+                            # The brackets might be empty for some functions...
+                            optional=True,
+                            ephemeral_name="FunctionContentsGrammar",
+                        ),
+                    )
+                ),
+            )
+        ),
+        Sequence(
+            Sequence(
+                AnyNumberOf(
```

Comment:
I presume the reason you are using AnyNumberOf with min and max of 1 is because it’s the only way you can exclude the DateAddFunctionNameSegment?

Reply:
@tunetheweb Yes, as far as I know.  If there's another way to perform the exclusion I can switch over.  I almost tried adding similar logic to Ref() but figured this would be less impactful.

Reply:
All good. Clever use if this!